File: src/main/java/net/frozenorb/potpvp/util/serialization/ItemStackAdapter.java
package net.frozenorb.potpvp.util.serialization;

import com.google.gson.*;
import net.frozenorb.potpvp.PotPvPRP;
import org.bukkit.Color;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.*;
import org.bukkit.potion.PotionEffect;

import java.lang.reflect.Type;
import java.util.*;
import java.util.Map.Entry;

@SuppressWarnings("deprecation")
public class ItemStackAdapter implements JsonDeserializer<ItemStack>,JsonSerializer<ItemStack> {

    public JsonElement serialize(ItemStack item, Type type, JsonSerializationContext context) {
        return ItemStackAdapter.serialize(item);
    }

    public ItemStack deserialize(JsonElement element, Type type, JsonDeserializationContext context) throws JsonParseException {
        return ItemStackAdapter.deserialize(element);
    }

    public static JsonElement serialize(ItemStack item) {
        if (item == null) {
            item=new ItemStack(Material.AIR);
        }
        final JsonObject element=new JsonObject();
        element.addProperty("id", item.getTypeId());
        element.addProperty(getDataKey(item), item.getDurability());
        element.addProperty("count", item.getAmount());
        if (item.hasItemMeta()) {
            final ItemMeta meta=item.getItemMeta();
            if (meta.hasDisplayName()) {
                element.addProperty("name", meta.getDisplayName());
            }
            if (meta.hasLore()) {
                element.add("lore", convertStringList(meta.getLore()));
            }
            if (meta instanceof LeatherArmorMeta) {
                element.addProperty("color", ((LeatherArmorMeta) meta).getColor().asRGB());
            } else if (meta instanceof SkullMeta) {
                element.addProperty("skull", ((SkullMeta) meta).getOwner());
            } else if (meta instanceof BookMeta) {
                element.addProperty("title", ((BookMeta) meta).getTitle());
                element.addProperty("author", ((BookMeta) meta).getAuthor());
                element.add("pages", convertStringList(((BookMeta) meta).getPages()));
            } else if (meta instanceof PotionMeta) {
                if (!((PotionMeta) meta).getCustomEffects().isEmpty()) {
                    element.add("potion-effects", convertPotionEffectList(((PotionMeta) meta).getCustomEffects()));
                }
            } else if (meta instanceof MapMeta) {
                element.addProperty("scaling", ((MapMeta) meta).isScaling());
            } else if (meta instanceof EnchantmentStorageMeta) {
                final JsonObject storedEnchantments=new JsonObject();
                for ( final Map.Entry<Enchantment, Integer> entry : ((EnchantmentStorageMeta) meta).getStoredEnchants().entrySet() ) {
                    storedEnchantments.addProperty(entry.getKey().getName(), entry.getValue());
                }
                element.add("stored-enchants", storedEnchantments);
            }
        }
        if (item.getEnchantments().size() != 0) {
            final JsonObject enchantments=new JsonObject();
            for ( final Map.Entry<Enchantment, Integer> entry2 : item.getEnchantments().entrySet() ) {
                enchantments.addProperty(entry2.getKey().getName(), entry2.getValue());
            }
            element.add("enchants", enchantments);
        }
        return element;
    }

    public static ItemStack deserialize(JsonElement object) {
        JsonObject enchantments;
        if (!(object instanceof JsonObject)) {
            return new ItemStack(Material.AIR);
        }
        JsonObject element=(JsonObject) object;
        int id=element.get("id").getAsInt();
        short data=element.has("damage") ? element.get("damage").getAsShort() : (element.has("data") ? element.get("data").getAsShort() : (short) 0);
        int count=element.get("count").getAsInt();
        ItemStack item=new ItemStack(id, count, data);
        ItemMeta meta=item.getItemMeta();
        if (element.has("name")) {
            meta.setDisplayName(element.get("name").getAsString());
        }
        if (element.has("lore")) {
            meta.setLore(ItemStackAdapter.convertStringList(element.get("lore")));
        }
        if (element.has("color")) {
            ((LeatherArmorMeta) meta).setColor(Color.fromRGB(element.get("color").getAsInt()));
        } else if (element.has("skull")) {
            ((SkullMeta) meta).setOwner(element.get("skull").getAsString());
        } else if (element.has("title")) {
            ((BookMeta) meta).setTitle(element.get("title").getAsString());
            ((BookMeta) meta).setAuthor(element.get("author").getAsString());
            ((BookMeta) meta).setPages(ItemStackAdapter.convertStringList(element.get("pages")));
        } else if (element.has("potion-effects")) {
            PotionMeta potionMeta=(PotionMeta) meta;
            for ( PotionEffect effect : ItemStackAdapter.convertPotionEffectList(element.get("potion-effects")) ) {
                potionMeta.addCustomEffect(effect, false);
            }
        } else if (element.has("scaling")) {
            ((MapMeta) meta).setScaling(element.get("scaling").getAsBoolean());
        } else if (element.has("stored-enchants")) {
            enchantments=(JsonObject) element.get("stored-enchants");
            for ( Enchantment enchantment : Enchantment.values() ) {
                if (!enchantments.has(enchantment.getName())) continue;
                ((EnchantmentStorageMeta) meta).addStoredEnchant(enchantment, enchantments.get(enchantment.getName()).getAsInt(), true);
            }
        }
        item.setItemMeta(meta);
        if (element.has("enchants")) {
            enchantments=(JsonObject) element.get("enchants");
            for ( Enchantment enchantment : Enchantment.values() ) {
                if (!enchantments.has(enchantment.getName())) continue;
                item.addUnsafeEnchantment(enchantment, enchantments.get(enchantment.getName()).getAsInt());
            }
        }
        return item;
    }

    private static String getDataKey(ItemStack item) {
        if (item.getType() == Material.AIR) {
            return "data";
        }
        if (Enchantment.DURABILITY.canEnchantItem(item)) {
            return "damage";
        }
        return "data";
    }

    public static JsonArray convertStringList(Collection<String> strings) {
        JsonArray ret=new JsonArray();
        for ( String string : strings ) {
            ret.add(new JsonPrimitive(string));
        }
        return ret;
    }

    public static List<String> convertStringList(JsonElement jsonElement) {
        JsonArray array=jsonElement.getAsJsonArray();
        ArrayList<String> ret=new ArrayList<>();
        for ( JsonElement element : array ) {
            ret.add(element.getAsString());
        }
        return ret;
    }

    public static JsonArray convertPotionEffectList(Collection<PotionEffect> potionEffects) {
        JsonArray ret=new JsonArray();
        for ( PotionEffect e : potionEffects ) {
            ret.add(PotionEffectAdapter.toJson(e));
        }
        return ret;
    }

    public static List<PotionEffect> convertPotionEffectList(JsonElement jsonElement) {
        if (jsonElement == null) {
            return null;
        }
        if (!jsonElement.isJsonArray()) {
            return null;
        }
        JsonArray array=jsonElement.getAsJsonArray();
        ArrayList<PotionEffect> ret=new ArrayList<>();
        for ( JsonElement element : array ) {
            PotionEffect e=PotionEffectAdapter.fromJson(element);
            if (e == null) continue;
            ret.add(e);
        }
        return ret;
    }
}

File: src/main/kotlin/net/frozenorb/potpvp/kt/util/serialization/ItemStackAdapter.kt
package net.frozenorb.potpvp.kt.util.serialization

import com.google.gson.*
import org.bukkit.Color
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.*
import org.bukkit.potion.PotionEffect
import java.lang.reflect.Type
import java.util.*

@Suppress
class ItemStackAdapter : JsonDeserializer<ItemStack>, JsonSerializer<ItemStack> {
    override fun serialize(item: ItemStack, type: Type, context: JsonSerializationContext): JsonElement {
        return serialize(item)
    }

    @Throws(JsonParseException::class)
    override fun deserialize(element: JsonElement, type: Type, context: JsonDeserializationContext): ItemStack {
        return deserialize(element)
    }

    companion object {

        fun serialize(item: ItemStack?): JsonElement {
            var item = item
            if (item == null) {
                item = ItemStack(Material.AIR)
            }

            val element = JsonObject()
            element.addProperty("id", item.typeId as Number)
            element.addProperty(getDataKey(item), item.durability as Number)
            element.addProperty("count", item.amount as Number)

            if (item.hasItemMeta()) {
                val meta = item.itemMeta
                if (meta.hasDisplayName()) {
                    element.addProperty("name", meta.displayName)
                }

                if (meta.hasLore()) {
                    element.add("lore", convertStringList(meta.lore) as JsonElement)
                }

                if (meta is LeatherArmorMeta) {
                    element.addProperty("color", meta.color.asRGB() as Number)
                } else if (meta is SkullMeta) {
                    element.addProperty("skull", meta.owner)
                } else if (meta is BookMeta) {
                    element.addProperty("title", meta.title)
                    element.addProperty("author", meta.author)
                    element.add("pages", convertStringList(meta.pages) as JsonElement)
                } else if (meta is PotionMeta) {
                    if (meta.customEffects.isNotEmpty()) {
                        element.add("potion-effects", convertPotionEffectList(meta.customEffects) as JsonElement)
                    }
                } else if (meta is MapMeta) {
                    element.addProperty("scaling", java.lang.Boolean.valueOf(meta.isScaling))
                } else if (meta is EnchantmentStorageMeta) {
                    val storedEnchantments = JsonObject()
                    for ((key, value) in meta.storedEnchants) {
                        storedEnchantments.addProperty(key.name, value as Number)
                    }
                    element.add("stored-enchants", storedEnchantments as JsonElement)
                }
            }

            if (item.enchantments.isNotEmpty()) {
                val enchantments = JsonObject()
                for (entry2 in item.enchantments.entries) {
                    enchantments.addProperty(entry2.key.name, entry2.value as Number)
                }
                element.add("enchants", enchantments as JsonElement)
            }

            return element
        }

        fun deserialize(`object`: JsonElement?): ItemStack {
            if (`object` == null || `object` !is JsonObject) {
                return ItemStack(Material.AIR)
            }
            val element = `object` as JsonObject?
            val id = element!!.get("id").asInt
            val data = (if (element.has("damage")) element.get("damage").asShort else if (element.has("data")) element.get("data").asShort else 0).toShort()
            val count = element.get("count").asInt
            val item = ItemStack(id, count, data)
            val meta = item.itemMeta

            if (element.has("name")) {
                meta.displayName = element.get("name").asString
            }

            if (element.has("lore")) {
                meta.lore = convertStringList(element.get("lore"))
            }

            if (element.has("color")) {
                (meta as LeatherArmorMeta).color = Color.fromRGB(element.get("color").asInt)
            } else if (element.has("skull")) {
                (meta as SkullMeta).owner = element.get("skull").asString
            } else if (element.has("title")) {
                (meta as BookMeta).title = element.get("title").asString
                meta.author = element.get("author").asString
                meta.pages = convertStringList(element.get("pages"))
            } else if (element.has("potion-effects")) {
                val potionMeta = meta as PotionMeta
                for (effect in convertPotionEffectList(element.get("potion-effects"))!!) {
                    potionMeta.addCustomEffect(effect, false)
                }
            } else if (element.has("scaling")) {
                (meta as MapMeta).isScaling = element.get("scaling").asBoolean
            } else if (element.has("stored-enchants")) {
                val enchantments = element.get("stored-enchants") as JsonObject
                for (enchantment in Enchantment.values()) {
                    if (enchantments.has(enchantment.name)) {
                        (meta as EnchantmentStorageMeta).addStoredEnchant(enchantment, enchantments.get(enchantment.name).asInt, true)
                    }
                }
            }

            item.itemMeta = meta

            if (element.has("enchants")) {
                val enchantments = element.get("enchants") as JsonObject
                for (enchantment in Enchantment.values()) {
                    if (enchantments.has(enchantment.name)) {
                        item.addUnsafeEnchantment(enchantment, enchantments.get(enchantment.name).asInt)
                    }
                }
            }

            return item
        }

        private fun getDataKey(item: ItemStack): String {
            if (item.type === Material.AIR) {
                return "data"
            }
            return if (Enchantment.DURABILITY.canEnchantItem(item)) {
                "damage"
            } else "data"
        }

        fun convertStringList(strings: Collection<String>): JsonArray {
            val ret = JsonArray()
            for (string in strings) {
                ret.add(JsonPrimitive(string) as JsonElement)
            }
            return ret
        }

        fun convertStringList(jsonElement: JsonElement): List<String> {
            val array = jsonElement.asJsonArray
            val ret = ArrayList<String>()
            for (element in array) {
                ret.add(element.asString)
            }
            return ret
        }

        fun convertPotionEffectList(potionEffects: Collection<PotionEffect>): JsonArray {
            val ret = JsonArray()
            for (e in potionEffects) {
                ret.add(PotionEffectAdapter.toJson(e) as JsonElement)
            }
            return ret
        }

        fun convertPotionEffectList(jsonElement: JsonElement?): List<PotionEffect>? {
            if (jsonElement == null) {
                return null
            }

            if (!jsonElement.isJsonArray) {
                return null
            }

            val array = jsonElement.asJsonArray
            val ret = ArrayList<PotionEffect>()
            for (element in array) {
                val e = PotionEffectAdapter.fromJson(element) ?: continue
                ret.add(e)
            }

            return ret
        }
    }
}